extension/src/completion/autocompletionProvider.ts

```diff
@@ -171,30 +187,222 @@ function getMatchingWord(document: vscode.TextDocument, position: vscode.Positio
+                    this.previousItems.unshift(context);
+                } else if (context === "DEFUNCHILD") {
+                    context = "EXIT";
+                    this.tryAddSymbol(atom.symbol, AutoLispKind.Defun); // only collect defun name
+                } else if (!this.rooted && this.previousItems[0] && (this.previousItems[0] === "DEFUN" || this.previousItems[0] === "DEFUN-Q")) {                    
+                    this.rooted = true;                    
+                    this.tryAddSymbol(atom.symbol, AutoLispKind.Defun); // collected root (first) defun name
+                } else if (this.previousItems[1] && (this.previousItems[1] === "DEFUN" || this.previousItems[1] === "DEFUN-Q")) {
+                    context = ucAtom === "(" ? "ARGS" : ""; // this context tells us when we are collecting args or locals
+                    this.previousItems.unshift(ucAtom); // most likely a '(' but defuns aren't required to have a (arg / local) declaration space so it could be nil
+                } else if (this.previousItems[0] && this.previousItems[0] === "SETQ") {
+                    context = 2; // this context tells me I am looking for a variable declaration
+                    this.tryAddSymbol(atom.symbol, AutoLispKind.Global);
+                } else {
+                    this.previousItems.unshift(ucAtom);
+                }
+            }
+        });
+    }
+
+    load(document: vscode.TextDocument, position: vscode.Position): void {
+        // get the scope of this position  
+        let kwRegex: RegExp = /\(\s*(DEFUN|DEFUN-Q)\s+\S{1,99}(\s|\(|;)/gi;
+        let tmp = findContainers(document, position);
+        let pr1: ElementRange = tmp.containerBlockComment;
+        let pr2: ElementRange[] = tmp.containerParens;
+        let contents: string = "";
+        const containers: ElementRange[] = findContainers(document, position).containerParens;
+        for (let ci = 0; ci < containers.length; ci++) {
+            const entry: ElementRange = containers[ci];
+            const value: string = document.getText(new vscode.Range(document.positionAt(entry.startPos.offsetInDocument), document.positionAt(entry.endPos.offsetInDocument)));
+            // retrieve the string contents of the first Defun scope
+            if (value.search(kwRegex) !== -1){                
+                contents = value;
+                break;
+            }
+        }
+        // Fallback for when the typed values are not in a Defun
+        if (contents === ""){
+            contents = document.getText(new vscode.Range(document.positionAt(containers.slice(-1)[0].startPos.offsetInDocument), document.positionAt(containers.slice(-1)[0].endPos.offsetInDocument)));
+        }
+        
+        
+        // Create and load that scope into a temporary document for contextual parsing, had to add this openFromContent method to narrow the LispParser SOW
```

Comment:
I think we can parse the whole file, don't need to find the scope container firstly. Please see the class `LispParser`